### PR TITLE
ICVL

### PR DESCRIPTION
**Pull request by chrsunwil**
Made the code much more general. Now uses the main_file.py to call functions from other files. Each dataset gets it's own file which implements the abstract class form DataSetInfoAbstractClass. This will ensure that each data set has a method for loading and visualizing. These get called in both the main_file.py and the shared_vae_class.py. The model is now also representing in a .png file saved into an output folder. The visualizations are also saved into this folder. I also made the ICVL and MNIST data into pickle files to speed up the process of loading in the data.

Adding new datasets should be a much easier process now as it will mainly involved adding a new file which maintains the invariants of the abstract class.

### CODE REVIEW

File: DataSetInfoAbstractClass.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+class dataSetInfoAbstract(object):
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def load(self):
+        return
+
+    @abstractmethod
+    def visualize(self, randIndexes, rightDomain, right_decoded_imgs, rightToLeftCycle, right_generatedImgs, leftToRightImgs,
```
[the remark this answers is not in the document]
Comment (chrsunwil) [reply DataSetInfoAbstractClass.py]: I was using https://packagecontrol.io/packages/Python%20PEP8%20Autoformat but it seems like it isn't the most reliable.

File: main_file.py [hunk 2]
```diff
@@ -1,30 +1,23 @@
...
-
-# Making Copies of the Data to creat Inverses
-import copy
+# MNIST 28x28
+# ICVL 60x80
 
-a_train = copy.copy(x_train)
-for i in range(len(a_train)):
-    a_train[i] = 1 - a_train[i]
+dataSetInfo = ICVL.dataInfo()
```
[the remark this answers is not in the document]
Comment (chrsunwil) [reply main_file.py]: I'm not sure if I can come up with a benefit which is exclusive with giving them specific names. It made more sense to me because they're implementations of the abstract function, so by using the same name they can interface really easily with only having to change which file they're called from. I figured that this would allow for a dictionary down the road which would mean that the user could input a dataset when they run the program, and not have to change anything in the files at all to switch between datasets. This, however, is basically just as easy to do with specific names. I originally gave them specific names but changed it. Basically, it just made the most sense to me in my head to emphasize that the functions interface the same way.

File: ICVL.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+
+class dataInfo(dataSetInfoAbstract):
+
+    def __init__(self):
+        self.name = "ICVL"
+
+    def load(self):
+
+        with open("Data/ICVL_Data/Training/ICVL_Training.pkl", "rb") as openfileobject:
```
[the remark this answers is not in the document]
Comment (chrsunwil) [reply ICVL.py]: I'm not sure if I necessarily agree with making this a function parameter. This is a very specific implementation and specifically meant for the ICVL data. Because I'm going to write a script that will download, clean, and place it in a set location, the only reason this would have to be a parameter is if they move the folders around. If they want to use their own data, then they'll create their own implementation of the abstract class which will handle the loading of data there.

Another way to approach this would be to make the loading of data generally and not in the abstract function, but I'm not sure if I see an advantage in that because of how specifically most of the data needs to be treated. I originally processed all my data each time inside the load function (Before I pickled them), and if the user wants to add their own data, that will be very unique to their application.

File: ICVL.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        return (x_train, a_train, x_test, a_test)
+
+    # This file plots the lines for the various fingers. This is all hardcoded
+    # from the files
+    def draw_hands(self, Xs, Ys):
+        plt.scatter(Xs, Ys)
+
+        plt.gca().invert_yaxis()
+
+        linesStart = np.array([Xs[0], Xs[1]])
```
[the remark this answers is not in the document]
Comment (chrsunwil) [reply ICVL.py]: You could probably shorten it by making several `for` loops and using arrays with the indices inside, but you wouldn't shorten it by very much (Because of specific colorings). If you were to implement a condition for checking which color to make it, I think you would end up with a somewhat complicated `for` loop. Because of the nature of the data's layout, there's not a particularly elegant way of addressing this, so far as I can see. I think that right now it's more readable (albeit ugly) than if you were to make a more complex, shorter way to do it.

**Comment by chrsunwil** [comment]
@gwaygenomics 

 - I added docstrings to all files and function headers.
 - I added the labels to the figures as per your link, work's reasonably well, but I had to put it on top, rather than the side of the image.
 - I changed the order of imports to what is in the document you linked.
 - I changed to use the `os.path.join` and Python3 string concatenation.
 - I added a parameter for number of visualizations to be shown
 - I used a linter to have proper formatting (I think)

I wrote some comments in response throughout your review.

File: ICVL.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply ICVL.py]: gotcha - ok, lets make some specific modifications then with that plan in mind.

How setting it to `self.training_file` in the `__init__` method? Also, should move towards the `os.path.join()` convention. This way a user can know where the file exists and it can be used later. Can then do `with open(self.training_file, 'rb') as openfileobject:...`

[the remark this answers is not in the document]
Comment (gwaygenomics) [reply ICVL.py]: can do for `self.testing_file` too

File: main_file.py [hunk 5]
```diff
@@ -1,30 +1,59 @@
-from shared_vae_class import shared_vae_class
-from model_objects import model_parameters
+"""
+shared-latent-space/main_file.py
 
+This file is the main file which should be run from command line
 
-import keras
-from keras.datasets import mnist
+Usage: Run ij comand line with no arguments:
```
Comment (gwaygenomics) [inline main_file.py]: typo

File: main_file.py [hunk 6]
```diff
@@ -1,30 +1,59 @@
...
+Output:
+    Represenation of the model's encoder saved to /Output/<dataset>
+        with the parameters of the model in the file name.
+    Visualiazations as defined in <dataset>'s corresponding implementation
+        of DataSetInfoAbstractClass. These may also be saved to the
+        /Output/<dataset> foled depending on the implementation.
+
+Currently, <dataset> can be ICVL or MNIST as included with this repo.
+
+TO DO: Make parameters into command line parameters.
```
Comment (gwaygenomics) [inline main_file.py]: TODOs belong in issues, not in committed code

Comment (gwaygenomics) [reply main_file.py]: Issues can also be linked to code with `#` + number

File: shared_vae_class.py [hunk 7]
```diff
@@ -35,7 +68,8 @@ def sampling(args):
             return z_mean + K.exp(z_log_sigma) * epsilon
 
         # Loss function for the VAE
-        # Loss function comprised of two parts, Cross_entropy, and divergense
+        # Loss function comprised of two parts, Cross_entropy, and
+        # divergense
```
Comment (gwaygenomics) [inline shared_vae_class.py]: typo

Comment (gwaygenomics) [reply shared_vae_class.py]: and a couple copy+pastes below

File: DataSetInfoAbstractClass.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+    @abstractmethod
+    def load(self):
+        """
+        Loads the testing and training data from pickle files.
+        This needs to return the specified data.
+
+        Args: None
+
+        Returns: (Float array, Float array, Float array, Float array)
```
[the remark this answers is not in the document]
Comment (chrsunwil) [reply DataSetInfoAbstractClass.py]: Well I didn't see a way to define the return types for an abstract class. From what I read, it seems like Python's abstract classes aren't really a full abstract class because they don't do any type checking. This seemed like a way to inform the reader of what the function should return in all implementations.

**Comment by chrsunwil** [comment]
@gwaygenomics  I have addressed your comments. My repo didn't have any .pyc files in it, so let me know if shared_vae_class.pyc is still there.
